cpp/ql/src/Likely Bugs/Underspecified Functions/MistypedFunctionArguments.ql

```diff
@@ -44,16 +40,23 @@ pragma[inline]
 predicate argTypeMayBeUsed(Type arg, Type parm) {
   arg = parm
   or
-  // we treat signed and unsigned versions of integer types as compatible.
+  // float arguments will have been promoted to double,
+  // and the parameter must match this
+  arg instanceof DoubleType and
+  parm instanceof DoubleType
+  or
+  // integral arguments are promoted to int (but not long long).
   arg instanceof IntegralType and
-  parm instanceof IntegralType
+  arg.getSize() = 4 and
+  parm instanceof IntegralType and
+  parm.getSize() = 4
   or
   // pointers to compatible types
   pointerArgTypeMayBeUsed(arg.(PointerType).getBaseType().getUnspecifiedType(),
-    parm.(PointerType).getBaseType().getUnspecifiedType())
+    parm.(PointerType).getBaseType().getUnderlyingType().getUnspecifiedType())
   or
   pointerArgTypeMayBeUsed(arg.(PointerType).getBaseType().getUnspecifiedType(),
-    parm.(ArrayType).getBaseType().getUnspecifiedType())
+    parm.(ArrayType).getBaseType().getUnderlyingType().getUnspecifiedType())
```

Comment:
If `.getUnderlyingType().getUnspecifiedType()` doesn't produce the same result as `.getUnspecifiedType()`, then I believe it's a bug. Do you have an example of that?

Reply:
I wasn't sure if they were the same which is why I used <code>.getUnderlyingType().getUnspecifiedType()</code> initially, but have since convinced myself that this is overkill.